data/transactions/logic/pairing.go

```diff
@@ -17,99 +17,876 @@
-	ret.X = bytesToBN254Field(b[:32])
-	ret.Y = bytesToBN254Field(b[32:64])
-	return
-}
+func (s sError) Error() string { return string(s) }
+
+const (
+	errNotOnCurve    = sError("point not on curve")
+	errWrongSubgroup = sError("wrong subgroup")
+	errEmptyInput    = sError("empty input")
+)
 
-func bytesToBN254G1s(b []byte) (ret []bn254.G1Affine) {
-	for i := 0; i < len(b)/64; i++ {
-		ret = append(ret, bytesToBN254G1(b[(i*64):(i*64+64)]))
+// Input: Two byte slices at top of stack, each an uncompressed point
+// Output: Single byte slice on top of stack which is the uncompressed sum of inputs
+func opEcAdd(cx *EvalContext) error {
+	group := EcGroup(cx.program[cx.pc+1])
+	fs, ok := ecGroupSpecByField(group)
+	if !ok { // no version check yet, both appeared at once
+		return fmt.Errorf("invalid ec_add group %s", group)
 	}
-	return
-}
 
-func bytesToBN254G2(b []byte) (ret bn254.G2Affine) {
-	ret.X.A0 = bytesToBN254Field(b[:32])
-	ret.X.A1 = bytesToBN254Field(b[32:64])
-	ret.Y.A0 = bytesToBN254Field(b[64:96])
-	ret.Y.A1 = bytesToBN254Field(b[96:128])
-	return
-}
+	last := len(cx.Stack) - 1
+	prev := last - 1
+	a := cx.Stack[prev].Bytes
+	b := cx.Stack[last].Bytes
 
-func bytesToBN254G2s(b []byte) (ret []bn254.G2Affine) {
-	for i := 0; i < len(b)/128; i++ {
-		ret = append(ret, bytesToBN254G2(b[(i*128):(i*128+128)]))
+	var res []byte
+	var err error
+	switch fs.field {
+	case BN254g1:
+		res, err = bn254G1Add(a, b)
+	case BN254g2:
+		res, err = bn254G2Add(a, b)
+	case BLS12_381g1:
+		res, err = bls12381G1Add(a, b)
+	case BLS12_381g2:
+		res, err = bls12381G2Add(a, b)
+	default:
+		err = fmt.Errorf("invalid ec_add group %s", group)
 	}
-	return
+	cx.Stack[prev].Bytes = res
+	cx.Stack = cx.Stack[:last]
+	return err
 }
 
-func bN254G1ToBytes(g1 *bn254.G1Affine) (ret []byte) {
-	retX := g1.X.Bytes()
-	retY := g1.Y.Bytes()
-	ret = append(retX[:], retY[:]...)
-	return
-}
+// Input: ToS is a scalar, encoded as an unsigned big-endian, second to top is
+// uncompressed bytes for g1 point
+// Output: Single byte slice on top of stack which contains uncompressed bytes
+// for product of scalar and point
+func opEcScalarMul(cx *EvalContext) error {
+	group := EcGroup(cx.program[cx.pc+1])
+	fs, ok := ecGroupSpecByField(group)
+	if !ok { // no version check yet, both appeared at once
+		return fmt.Errorf("invalid ec_scalar_mul group %s", group)
+	}
 
-func opBn256Add(cx *EvalContext) error {
 	last := len(cx.Stack) - 1
 	prev := last - 1
 	aBytes := cx.Stack[prev].Bytes
-	bBytes := cx.Stack[last].Bytes
-	if len(aBytes) != 64 || len(bBytes) != 64 {
-		return errors.New("expect G1 in 64 bytes")
-	}
-	a := bytesToBN254G1(aBytes)
-	b := bytesToBN254G1(bBytes)
-	res := new(bn254.G1Affine).Add(&a, &b)
-	resBytes := bN254G1ToBytes(res)
+	kBytes := cx.Stack[last].Bytes
+	if len(kBytes) > scalarSize {
+		return fmt.Errorf("ec_scalar_mul scalar len is %d, exceeds 32", len(kBytes))
+	}
+	k := new(big.Int).SetBytes(kBytes)
```

Comment:
I think I once convinced myself that the second form is just as bad, because it's really calling a function like: `SetBytes(&big, b)`, and the Go compiler therefore assumes that `&big` escapes, and must be heap allocated, even though you don't see it. So it's not that the `new` gets optimized away, it's that the declaration on the stack gets "pessimized" up to an allocation.

I will write a benchmark to convince myself.

Reply:
Yes, it gets pessimized. I ran my benchmark for `ec_scalar_mul` with `b.ReportAllocs()` and found that there are 13 allocations per loop regardless of how it's written.